src/network/network_gui.cpp

```diff
@@ -1564,333 +1575,735 @@ NetworkCompanyInfo *GetLobbyCompanyInfo(CompanyID company)
 	{
-		Dimension d = *size;
-		for (const ClientListAction &action : this->actions) {
-			d = maxdim(GetStringBoundingBox(action.name), d);
+		switch (widget) {
+			case WID_CL_SERVER_VISIBILITY:
+				*size = maxdim(GetStringBoundingBox(_server_visibility_dropdown[0]), GetStringBoundingBox(_server_visibility_dropdown[1]));
+				size->width += padding.width;
+				size->height += padding.height;
+				break;
+
+			case WID_CL_MATRIX: {
+				uint height = std::max({GetSpriteSize(SPR_COMPANY_ICON).height, GetSpriteSize(SPR_JOIN).height, GetSpriteSize(SPR_ADMIN).height, GetSpriteSize(SPR_CHAT).height});
+				height += ScaleGUITrad(WD_FRAMERECT_TOP) + ScaleGUITrad(WD_FRAMERECT_BOTTOM);
+				this->line_height = std::max(height, (uint)FONT_HEIGHT_NORMAL) + ScaleGUITrad(WD_MATRIX_TOP + WD_MATRIX_BOTTOM);
+
+				resize->width = 1;
+				resize->height = this->line_height;
+				fill->height = this->line_height;
+				size->height = std::max(size->height, 5 * this->line_height);
+				break;
+			}
 		}
+	}
 
-		d.height *= (uint)this->actions.size();
-		d.width += WD_FRAMERECT_LEFT + WD_FRAMERECT_RIGHT;
-		d.height += WD_FRAMERECT_TOP + WD_FRAMERECT_BOTTOM;
-		*size = d;
+	void OnResize() override
+	{
+		this->vscroll->SetCapacityFromWidget(this, WID_CL_MATRIX);
 	}
 
-	void DrawWidget(const Rect &r, int widget) const override
+	void SetStringParameters(int widget) const override
 	{
-		/* Draw the actions */
-		int sel = this->sel_index;
-		int y = r.top + WD_FRAMERECT_TOP;
-		for (const ClientListAction &action : this->actions) {
-			TextColour colour;
-			if (sel-- == 0) { // Selected item, highlight it
-				GfxFillRect(r.left + 1, y, r.right - 1, y + FONT_HEIGHT_NORMAL - 1, PC_BLACK);
-				colour = TC_WHITE;
-			} else {
-				colour = TC_BLACK;
-			}
+		switch (widget) {
+			case WID_CL_SERVER_NAME:
+				SetDParamStr(0, _settings_client.network.server_name);
+				break;
 
-			DrawString(r.left + WD_FRAMERECT_LEFT, r.right - WD_FRAMERECT_RIGHT, y, action.name, colour);
-			y += FONT_HEIGHT_NORMAL;
+			case WID_CL_SERVER_VISIBILITY:
+				SetDParam(0, _server_visibility_dropdown[_settings_client.network.server_advertise]);
+				break;
+
+			case WID_CL_CLIENT_NAME:
+				SetDParamStr(0, _settings_client.network.client_name);
+				break;
 		}
 	}
 
-	void OnMouseLoop() override
+	void OnClick(Point pt, int widget, int click_count) override
 	{
-		/* We selected an action */
-		uint index = (_cursor.pos.y - this->top - WD_FRAMERECT_TOP) / FONT_HEIGHT_NORMAL;
+		switch (widget) {
+			case WID_CL_SERVER_NAME_EDIT:
+				if (!_network_server) break;
 
-		if (_left_button_down) {
-			if (index == this->sel_index || index >= this->actions.size()) return;
+				this->query_widget = WID_CL_SERVER_NAME_EDIT;
+				SetDParamStr(0, _settings_client.network.server_name);
+				ShowQueryString(STR_JUST_RAW_STRING, STR_NETWORK_CLIENT_LIST_SERVER_NAME_QUERY_CAPTION, NETWORK_NAME_LENGTH, this, CS_ALPHANUMERAL, QSF_ENABLE_DEFAULT | QSF_LEN_IN_CHARS);
```

Comment:
I'd remove the option for a "default" server name, i.e. remove the `QSF_ENABLE_DEFAULT | `.